File: python/infinicore/ops/asum.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def asum(x: Tensor, *, out=None):
    if out is None:
        return Tensor(_infinicore.asum(x._underlying))

    _infinicore.asum_(x._underlying, out._underlying)

    return out

File: src/infinicore/ops/asum/asum.cc
#include "infinicore/ops/asum.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

common::OpDispatcher<Asum::schema> &Asum::dispatcher() {
    static common::OpDispatcher<Asum::schema> dispatcher_;
    return dispatcher_;
};

void Asum::execute(Tensor result, Tensor x) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(result, x);
    infinicore::context::setDevice(result->device());
    dispatcher().lookup(result->device().getType())(result, x);
}

Tensor asum(Tensor x) {
    auto result = Tensor::empty({}, x->dtype(), x->device());
    asum_(result, x);
    return result;
}

void asum_(Tensor result, Tensor x) {
    Asum::execute(result, x);
}

} // namespace infinicore::op
